src/main/java/com/znsio/teswiz/runner/DeviceSetup.java

```diff
@@ -82,17 +83,68 @@ static ArrayList<String> setupAndroidExecution() {
     static void verifyAppExistsAtMentionedPath() {
         String appPath = Setup.getFromConfigs(APP_PATH);
         LOGGER.info(String.format("Update path to Apk: %s", appPath));
-        if(appPath.equals(NOT_SET)) {
+        if (appPath.equals(NOT_SET)) {
             appPath = getAppPathFromCapabilities();
+            appPath = convertAppPathToFilePathIfNeeded(appPath);
             Setup.addToConfigs(APP_PATH, appPath);
-            String capabilitiesFileName = Setup.getFromConfigs(CAPS);
-            checkIfAppExistsAtTheMentionedPath(appPath, capabilitiesFileName);
         } else {
+            appPath = convertAppPathToFilePathIfNeeded(appPath);
             LOGGER.info(String.format("\tUsing AppPath provided as environment variable -  %s",
-                                      appPath));
+                    appPath));
         }
     }
 
+    private static String convertAppPathToFilePathIfNeeded(String appPath){
+        if (isAppPathAUrl(appPath)) {
+            String directoryPath = System.getProperty("user.dir") + File.separator + "temp" + File.separator + "sampleApps";
+            String fileName = appPath.split(File.separator)[appPath.split(File.separator).length - 1];
+            String filePath = directoryPath + File.separator + fileName;
+            createDirectoryIfNotExist(directoryPath);
+            downloadFileIfNotExist(appPath, filePath);
+            LOGGER.info("Changing value of appPath from URL to file path");
+            LOGGER.info("Before change, appPath value: "+ appPath);
+            appPath = filePath;
+            LOGGER.info("After change, appPath value: "+ filePath);
+        }
+        try{
+            new FileInputStream(appPath);
+        } catch (FileNotFoundException e) {
+            throw new RuntimeException("File not found: " + appPath, e);
+        }
+        return appPath;
+    }
+    private static void createDirectoryIfNotExist(String directoryPath){
+        try {
+            Path pathForDirectory = Path.of(directoryPath);
+            if(Files.exists(pathForDirectory))
+                LOGGER.info("Directory already exist");
+            else{
+                Files.createDirectories(pathForDirectory);
+                LOGGER.info("Creating directory");
+            }
+        } catch (IOException e) {
+        throw new RuntimeException("Failed to create directory: "+ directoryPath + ", error occurred" + e);
+        }
+    }
+
+    private static void downloadFileIfNotExist(String url, String filePath) {
+        File file = new File(filePath);
+        if (!file.exists()) {
+            LOGGER.info("File doesn't exist, need to download it");
+            try {
+                URL fileUrl = new URL(url);
```

Comment:
We have removed the checks for Malformed Urls and they are not handled too. If this method throws any exception for URL we are not handling it.